src/Modules/OrchardCore.Commerce/Services/ProductListService.cs

```diff
@@ -0,0 +1,100 @@
+using YesSql;
+using YesSql.Services;
+
+namespace OrchardCore.Commerce.Services;
+
+public class ProductListService : IProductListService
+{
+    private readonly ISession _session;
+    private readonly IEnumerable<IProductListFilterProvider> _productListQueryProviders;
+    private readonly IContentDefinitionManager _contentDefinitionManager;
+
+    public ProductListService(
+        ISession session,
+        IEnumerable<IProductListFilterProvider> productListQueryProviders,
+        IContentDefinitionManager contentDefinitionManager)
+    {
+        _session = session;
+        _productListQueryProviders = productListQueryProviders;
+        _contentDefinitionManager = contentDefinitionManager;
+    }
+
+    public async Task<ProductList> GetProductsAsync(ProductListPart productList, ProductListFilterParameters filterParameters)
+    {
+        ArgumentNullException.ThrowIfNull(productList);
+        ArgumentNullException.ThrowIfNull(filterParameters);
+
+        var productTypes = _contentDefinitionManager.ListTypeDefinitions()
+            .Where(type => type.Parts.Any(part => part.PartDefinition.Name == nameof(ProductPart)))
+            .Select(type => type.Name)
+            .ToArray();
+
+        var applicableProviders = await GetOrderedApplicableProvidersAsync(productList);
+
+        var query = _session.Query<ContentItem>();
+        query = query.With<ContentItemIndex>(index => index.ContentType.IsIn(productTypes) && index.Published);
+
+        var context = new ProductListFilterContext
+        {
+            ProductList = productList,
+            FilterParameters = filterParameters,
+            Query = query,
+        };
+
+        if (string.IsNullOrEmpty(filterParameters.OrderBy))
+        {
+            filterParameters.OrderBy = ProductListTitleFilterProvider.TitleAscOrderById;
+        }
+
+        foreach (var provider in applicableProviders)
+        {
+            context.Query = await provider.BuildQueryAsync(context) ?? context.Query;
+        }
+
+        var totalItemCount = await query.CountAsync();
+        var contentItems = await query.PaginateAsync(filterParameters.Pager.Page - 1, filterParameters.Pager.PageSize);
+
+        return new ProductList
+        {
+            Products = contentItems,
+            TotalItemCount = totalItemCount,
+        };
+    }
+
+    public async Task<IEnumerable<string>> GetOrderByOptionsAsync(ProductListPart productList)
+    {
+        ArgumentNullException.ThrowIfNull(productList);
+
+        var applicableProviders = await GetOrderedApplicableProvidersAsync(productList);
+
+        return (await applicableProviders.AwaitEachAsync(provider => provider.GetOrderByOptionIdsAsync(productList)))
+            .SelectMany(options => options);
+    }
+
+    public async Task<IEnumerable<string>> GetFilterIdsAsync(ProductListPart productList)
+    {
+        if (productList is null)
+        {
+            throw new ArgumentNullException(nameof(productList));
+        }
```

Comment:
You forgot this one.
```suggestion
        ArgumentNullException.ThrowIfNull(productList);
```